lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs

```diff
@@ -291,18 +287,23 @@ getTokenTransfer = label "getTokenTransfer" $ do
         value <- Coin <$> getWord64be
         return $ TxOut addr value
 
-    getAddress = do
-        headerByte <- getWord8
-        let kind = kindValue headerByte
-        let _discrimination = discriminationValue headerByte
-        case kind of
-            -- Single Address
-            0x3 -> Address <$> getByteString 32
-            0x4 -> error "unimplemented group address decoder"
-            0x5 -> error "unimplemented account address decoder"
-            0x6 -> error "unimplemented multisig address decoder"
-            other -> fail $ "Invalid address type: " ++ show other
 
+
+getAddress :: Get Address
+getAddress = do
+    -- We use 'lookAhead' to not consume the header, and let it
+    -- be included in the underlying Address ByteString.
+    headerByte <- label "address header" . lookAhead $ getWord8
+    let kind = kindValue headerByte
```

Comment:
Also, @KtorZ, there is difference in style here with the binary arithmetic:

```haskell
    case kind of
        -- Single Address
        0x3 -> Address <$> getByteString 33
        0x4 -> error "unimplemented group address decoder"
        0x5 -> error "unimplemented account address decoder"
        0x6 -> error "unimplemented multisig address decoder"
        other -> fail $ "Invalid address type: " ++ show other
  where
    kindValue :: Word8 -> Word8
    kindValue = (.&. 0b01111111)

    discriminationValue :: Word8 -> Network
    discriminationValue b = case b .&. 0b10000000 of
        0 -> Mainnet
        _ -> Testnet
```

vs hard-coded single/grouped values specific to each network found in `Cardano.Wallet.Jormungandr.Environment`
```haskell
instance KnownNetwork 'Mainnet where
    networkVal = Mainnet
    hrp = unsafeHumanReadablePart "ca"
    single = 0x03
    grouped = 0x04

instance KnownNetwork 'Testnet where
    networkVal = Testnet
    hrp = unsafeHumanReadablePart "ta"
    single = 0x83
    grouped = 0x84
```

I'm not sure what should be done. Do we really want to throw the arithmetic away?

Reply:
It's not about throwing it away and it's not about using hard-coded bytes or extracting them from the binary data. It's about checking that both matches. That's the only reason to have a discriminant version, to make sure that we do not use addresses on the wrong network.

We do already perform this check as part of the encode/decode functions since this is our entry-point for user-facing data. We could also do it in the binary modules when parsing blocks although we do expect the underlying node to do that for us already.